pkg/cstor/pvc_operation.go

```diff
@@ -0,0 +1,167 @@
+func (p *Plugin) backupPVC(volumeID string) error {
+	vol := p.volumes[volumeID]
+	var bkpPvc *v1.PersistentVolumeClaim
+
+	pvcs, err := p.K8sClient.
+		CoreV1().
+		PersistentVolumeClaims(vol.namespace).
+		List(metav1.ListOptions{})
+	if err != nil {
+		p.Log.Errorf("Error fetching PVC list : %s", err.Error())
+		return errors.New("Failed to fetch PVC list")
+	}
+
+	for _, pvc := range pvcs.Items {
+		if pvc.Spec.VolumeName == vol.volname {
+			bkpPvc = &pvc
+			break
+		}
+	}
+
+	if bkpPvc == nil {
+		p.Log.Errorf("Failed to find PVC for PV{%s}", vol.volname)
+		return errors.Errorf("Failed to find PVC for volume{%s}", vol.volname)
+	}
+
+	bkpPvc.ResourceVersion = ""
+	bkpPvc.SelfLink = ""
+	bkpPvc.Annotations = nil
+	bkpPvc.UID = ""
+	bkpPvc.Spec.VolumeName = ""
+
+	data, err := json.MarshalIndent(bkpPvc, "", "\t")
+	if err != nil {
+		return errors.New("Error doing json parsing")
+	}
+
+	filename := p.cl.GenerateRemoteFilename(vol.volname, vol.backupName)
+	if filename == "" {
+		return errors.New("Error creating remote file name for pvc backup")
+	}
+
+	if ok := p.cl.Write(data, filename+".pvc"); !ok {
+		return errors.New("Failed to upload PVC")
+	}
+
+	return nil
+}
+
+// createPVC create PVC for given volume name
+func (p *Plugin) createPVC(volumeID, snapName string) (*Volume, error) {
+	var pvc v1.PersistentVolumeClaim
+	var data []byte
+	var ok bool
+
+	filename := p.cl.GenerateRemoteFilename(volumeID, snapName)
+	if filename == "" {
+		return nil, errors.New("Error creating remote file name for pvc backup")
+	}
+
+	if data, ok = p.cl.Read(filename + ".pvc"); !ok {
+		return nil, errors.New("Failed to download PVC")
+	}
+
+	if err := json.Unmarshal(data, &pvc); err != nil {
+		return nil, errors.New("Failed to decode pvc")
+	}
+
+	newVol, err := p.getVolumeFromPVC(pvc)
+	if err == nil {
+		newVol.backupName = snapName
+		return newVol, nil
+	}
+
+	pvc.Annotations = make(map[string]string)
+	pvc.Annotations["openebs.io/created-through"] = "restore"
+	rpvc, err := p.K8sClient.
+		CoreV1().
+		PersistentVolumeClaims(pvc.Namespace).
+		Create(&pvc)
+	if err != nil {
+		return nil, errors.Errorf("Failed to create PVC : %s", err.Error())
+	}
+
+	for {
+		pvc, err := p.K8sClient.
+			CoreV1().
+			PersistentVolumeClaims(rpvc.Namespace).
+			Get(rpvc.Name, metav1.GetOptions{})
+		if err != nil || pvc.Status.Phase == v1.ClaimLost {
+			if err := p.K8sClient.
+				CoreV1().
+				PersistentVolumeClaims(pvc.Namespace).
+				Delete(rpvc.Name, nil); err != nil {
+				p.Log.Warnf("Failed to delete pvc {%s} : %s", rpvc.Name, err.Error())
+			}
+			return nil, errors.Errorf("Failed to create PVC : %s", err.Error())
+		}
+		if pvc.Status.Phase == v1.ClaimBound {
+			p.Log.Infof("PVC(%v) created..", pvc.Name)
+			vol := &Volume{
+				volname:    pvc.Spec.VolumeName,
+				namespace:  pvc.Namespace,
+				backupName: snapName,
+				casType:    *pvc.Spec.StorageClassName,
+			}
+
+			if err = p.waitForAllCVR(vol); err != nil {
```

Comment:
we can move this out of for loop as part of code reorg

Reply:
yes